### ISSUE TITLE
Allow to pass through custom driver options as part of 'minishift start'

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#305]
Some drivers support additional arguments to for example configure/enable host folder mounts or to select a switch (Hyper-V). AFAICT, there is no mechanism in Minishift atm which allows for passing custom arguments to the driver.

**Comment by hferentschik** [comment]
This has to be in some form generic, since we cannot add all the various options to the Minishift CLI. 

One thing I am considering is to drive this purely through persistent configuration and _config.json_. Given that we are targeting advanced use cases this might be a good trade-off. The idea would be to have a hypervisor specific section in the config, eg _hyper-v_ with key/value pairs of custom options. At startup we would then check whether for the selected driver there are any custom options configured. If so we would apply them.

**Comment by hferentschik** [comment]
Actually we can already pass through options. The way the drivers handle this is by taking parameters via environment variables. See also https://github.com/zchee/docker-machine-driver-xhyve/blob/master/xhyve/xhyve.go and https://github.com/docker/machine/blob/master/drivers/hyperv/hyperv.go. 

Not sure whether this is enough or whether we should put something around this. At the very least it needs documenting.

**Comment by gbraad** [comment]
@hferentschik have you been able to verify the use of the environment variables? I tried just now, and it seems they are ignored.

**Comment by hferentschik** [comment]
See https://github.com/minishift/minishift/issues/590#issuecomment-293397120

**Comment by hferentschik** [comment]
Using the algorithm described in issue #590, we should be able to pass any of the environment variables a driver x supports through to the actual driver. See for example https://docs.docker.com/machine/drivers/hyper-v/ for the Hyper-V options.

Obviously there is an overlap here with the options one can set via the command line or the config file. A priority order needs to be determined. If a user sets a driver specific environment variable, when does it get applied. Do we first apply the environment variables and then the settings from the command line or vice versa. 

We can also consider whether we want the user to prefix the variables with 'MINISHIFT_' This would avoid potential conflicts when users set these variables for the use with docker-machine as well. Also it might be more aligned with the current use of Minishift env variables. 

Last but not least, we can also discuss (potentially as a follow up) whether we want to allow setting the driver CLI option via config.json (and in turn via _config set_). So for example, one could also add 'hyperv-virtual-switch' to  the config. The benefit is that the user does not have to remember to set the env variable. Note, this does not mean that we add the option to our cli. We just allow it to be set. This implies we loosen the strict validation we have for _config set_ which might be a good idea anyways.

Last but not least, for the CLI we can add a generic '--driver-opts' flag which would be a comma separated list of key/value pairs specifying additional driver options. 

We can split this up into multiple steps/issues. First step is to generically determine the options of a driver and setting them.

**Comment by LalatenduMohanty** [comment]
> We can also consider whether we want the user to prefix the variables with 'MINISHIFT_' This would avoid potential conflicts when users set these variables for the use with docker-machine as well. Also it might be more aligned with the current use of Minishift env variables.

+1. I think this is a good idea.

**Comment by budhrg** [comment]
Hi @hferentschik , 

I did try to use reflection here to set the fields of driver for kvm, check below
```
func createKVMHost(config MachineConfig) (drivers.Driver, error) {
  api := libmachine.NewClient("", "")
  host, _ := api.NewHost("kvm", []byte("{}"))

  driver := host.Driver

  fmt.Println("Type of driver: ", reflect.TypeOf(driver))

  // Using  http://stackoverflow.com/a/6396678/1120530
  s := reflect.Indirect(reflect.ValueOf(driver))

  fmt.Println("Kind of s : ", s.Kind())
  
  if s.Kind() == reflect.Struct {
    f := s.FieldByName("Memory")
    
    if f.IsValid() {          #### f is invalid here since there is no Field like Memory in rpc.RPCClientDriver
      if f.CanSet() {
        if f.Kind() == reflect.Int {
          x := int64(config.Memory)
          if !f.OverflowInt(x) {
            f.SetInt(x)
          }
        }
      }
    }
  }

  fmt.Println("driver memory : ", s.FieldByName("Memory").String())

  fmt.Println("Type of driver: ", reflect.TypeOf(driver))

  if err := setDriverOptionsFromEnvironment(driver); err != nil {
    return nil, err
  }

  return driver, nil
}
```

It seems the `host.Driver` return the driver of type `rpc.RPCClientDriver` which doesn't have any field like `Memory`, or else

```
Starting local OpenShift cluster using 'kvm' hypervisor...
...
(default) Calling .SetConfigFromFlags
E0413 23:40:41.612434     482 start.go:168] Error starting the VM: gob: type not registered for interface: drivers.CheckDriverOptions. Retrying.
Found binary path at /usr/local/bin/docker-machine-driver-kvm
Launching plugin server for driver kvm

Type of driver:  *rpcdriver.RPCClientDriver
Kind of s :  struct
driver memory :  <invalid Value>
Type of driver:  *rpcdriver.RPCClientDriver

Plugin server listening at address 127.0.0.1:39403
() Calling .GetVersion
Using API Version  1
() Calling .SetConfigRaw
() Calling .GetMachineName
(default) Calling .GetMachineName
(default) Calling .DriverName
(default) Calling .GetCreateFlags

Type of driver:  *rpcdriver.RPCClientDriver
Kind of s :  struct
driver memory :  <invalid Value>
Type of driver:  *rpcdriver.RPCClientDriver

(default) Calling .SetConfigFromFlags
E0413 23:40:41.642119     482 start.go:168] Error starting the VM: gob: type not registered for interface: drivers.CheckDriverOptions. Retrying.
...
```

Let me know if I am doing anything wrong here.

**Comment by budhrg** [comment]
I did big digging into docker/machine code and found how docker-machine is able to use kvm driver.

Here (https://github.com/docker/machine/blob/master/commands/create.go#L324-L340), they are adding the flags supported by kvm to the `create` command dynamically and using it.

I am now investigating on the same approach.

**Comment by hferentschik** [comment]
Merged via pull request #773
